wagtail/contrib/redirects/signal_handlers.py

```diff
@@ -0,0 +1,166 @@
+import logging
+
+from typing import Iterable, Set, Tuple
+
+from django.conf import settings
+from django.db.models import Q
+
+from wagtail.core.models import Page, Site
+from wagtail.core.utils import BatchCreator, get_dummy_request
+
+from .models import Redirect
+
+
+logger = logging.getLogger(__name__)
+
+
+class BatchRedirectCreator(BatchCreator):
+    """
+    A specialized ``BatchCreator`` class for saving ``Redirect`` objects.
+    """
+
+    model = Redirect
+
+    def pre_process(self):
+        # delete any existing automatically-created redirects that might clash
+        # with the items in `self.items`
+        clashes_q = Q()
+        for item in self.items:
+            clashes_q |= Q(old_path=item.old_path, site_id=item.site_id)
+        Redirect.objects.filter(automatically_created=True).filter(clashes_q).delete()
+
+
+def autocreate_redirects_on_slug_change(
+    instance_before: Page,
+    instance: Page,
+    **kwargs
+):
+    if not getattr(settings, "WAGTAILREDIRECTS_AUTO_CREATE", True):
+        return None
+
+    # Determine sites to create redirects for
+    sites = Site.objects.filter(id__in=[
+        option.site_id for option in instance._get_relevant_site_root_paths(cache_object=instance)
+    ]).exclude(root_page=instance)
+
+    create_redirects(page=instance, page_old=instance_before, sites=sites)
+
+
+def autocreate_redirects_on_page_move(
+    instance: Page,
+    url_path_after: str,
+    url_path_before: str,
+    **kwargs,
+) -> None:
+    if not getattr(settings, "WAGTAILREDIRECTS_AUTO_CREATE", True):
+        return None
+
+    if url_path_after == url_path_before:
+        # Redirects are not needed for a page 'reorder'
+        return None
+
+    # NB: create_redirects expects specific instances
+    instance = instance.specific
```

Comment:
I will double check, but I believe the `page_slug_changed` signal provides specific pages already... I did things that way, because it would be very easy to do `old_instance.specific` in a handler and lose all of the original values.

Reply:
@gasman Have added a couple of comments in 1c47145 to help clarify this